In [1]:
import pandas as pd

Pandas is a large and powerful library.  Here we will only use a few of its basic features.  The main data structure that Pandas works with is called a "data frame".  This is a two-dimensional table of data in which the rows typically represent cases (e.g. NHANES subjects), and the columns represent variables.  Pandas also has a one-dimensional data structure called a `Series` that we will encounter occasionally.

Pandas has a variety of functions named with the pattern '`read_xxx`' for reading data in different formats into Python.  Right now we will focus on reading '`csv`' files, so we are using the '`read_csv`' function, which can read csv (and "tsv") format files that are exported from spreadsheet software like Excel.  The '`read_csv`' function by default expects the first row of the data file to contain column names.  

Using '`read_csv`' in its default mode is fairly straightforward.  There are many options to '`read_csv`' that are useful for handling less-common situations.  For example, you would use the option `sep='\t'` instead of the default `sep=','` if the fields of your data file are delimited by tabs instead of commas.  See [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) for the full documentation for '`read_csv`'.

Pandas can read a data file over the internet when provided with a URL, which is what we will do below.  In the Python script we will name the data set '`da`', i.e. this is the name of the Python variable that will hold the data frame after we have loaded it. 

The variable '`url`' holds a string (text) value, which is the internet URL where the data are located.  If you have the data file in your local filesystem, you can also use '`read_csv`' to read the data from this file.  In this case you would pass a file path instead of a URL, e.g. `pd.read_csv("my_file.csv")` would read a file named `my_file.csv` that is located in your current working directory.

In [18]:
url = "C:/Users/Administrator/Downloads/nhanes_2015_2016.csv"
da = pd.read_csv(url)
da

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0     83732     1.0     NaN     1.0       1         1        62         3   
1     83733     1.0     NaN     6.0       1         1        53         3   
2     83734     1.0     NaN     NaN       1         1        78         3   
3     83735     2.0     1.0     1.0       2         2        56         3   
4     83736     2.0     1.0     1.0       2         2        42         4   
...     ...     ...     ...     ...     ...       ...       ...       ...   
5730  93695     2.0     2.0     NaN       1         2        76         3   
5731  93696     2.0     2.0     NaN       2         1        26         3   
5732  93697     1.0     NaN     1.0       1         2        80         3   
5733  93700     NaN     NaN     NaN       1         1        35         3   
5734  93702     1.0     NaN     2.0       2         2        24         3   

      DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0          1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1          2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2          1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   
3          1.0       5.0  ...   134.0    68.0  109.8  160.9    42.4    38.5   
4          1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   
...        ...       ...  ...     ...     ...    ...    ...     ...     ...   
5730       1.0       3.0  ...   112.0    46.0   59.1  165.8    21.5    38.2   
5731       1.0       5.0  ...   116.0    76.0  112.1  182.2    33.8    43.4   
5732       1.0       4.0  ...   146.0    58.0   71.7  152.2    31.0    31.3   
5733       2.0       1.0  ...   106.0    66.0   78.2  173.3    26.0    40.3   
5734       1.0       5.0  ...   114.0    68.0   58.3  165.0    21.4    38.2   

      BMXARML  BMXARMC  BMXWAIST  HIQ210  
0        43.6     35.9     101.1     2.0  
1        40.0     33.2     107.9     NaN  
2        37.0     31.0     116.5     2.0  
3        37.7     38.3     110.1     2.0  
4        36.0     27.2      80.4     2.0  
...       ...      ...       ...     ...  
5730     37.0     29.5      95.0     2.0  
5731     41.8     42.3     110.2     2.0  
5732     37.5     28.8       NaN     2.0  
5733     37.5     30.6      98.9     2.0  
5734     33.5     26.2      72.5     2.0  

[5735 rows x 28 columns]

To confirm that we have actually obtained the data the we are expecting, we can display the shape (number of rows and columns) of the data frame in the notebook.  Note that the final expression in any Jupyter notebook cell is automatically printed, but you can force other expressions to be printed by using the '`print`' function, e.g. '`print(da.shape)`'.

Based on what we see below, the data set being read here has 5735 rows, corresponding to 5735 people in this wave of the NHANES study, and 28 columns, corresponding to 28 variables in this particular data file.  Note that NHANES collects thousands of variables on each study subject, but here we are working with a reduced file that contains a limited number of variables.

In [9]:
da.shape

(5735, 28)

### Exploring the contents of a data set

Pandas has a number of basic ways to understand what is in a data set.  For example, above we used the '`shape`' method to determine the numbers of rows and columns in a data set.  The columns in a Pandas data frame have names, to see the names, use the '`columns`' method:

In [10]:
da.columns

Index(['SEQN', 'ALQ101', 'ALQ110', 'ALQ130', 'SMQ020', 'RIAGENDR', 'RIDAGEYR',
       'RIDRETH1', 'DMDCITZN', 'DMDEDUC2', 'DMDMARTL', 'DMDHHSIZ', 'WTINT2YR',
       'SDMVPSU', 'SDMVSTRA', 'INDFMPIR', 'BPXSY1', 'BPXDI1', 'BPXSY2',
       'BPXDI2', 'BMXWT', 'BMXHT', 'BMXBMI', 'BMXLEG', 'BMXARML', 'BMXARMC',
       'BMXWAIST', 'HIQ210'],
      dtype='object')

Every variable in a Pandas data frame has a data type.  There are many different data types, but most commonly you will encounter floating point values (real numbers), integers, strings (text), and date/time values.  When Pandas reads a text/csv file, it guesses the data types based on what it sees in the first few rows of the data file.  Usually it selects an appropriate type, but occasionally it does not.  To confirm that the data types are consistent with what the variables represent, inspect the '`dtypes`' attribute of the data frame.

In [11]:
da.dtypes

SEQN          int64
ALQ101      float64
ALQ110      float64
ALQ130      float64
SMQ020        int64
RIAGENDR      int64
RIDAGEYR      int64
RIDRETH1      int64
DMDCITZN    float64
DMDEDUC2    float64
DMDMARTL    float64
DMDHHSIZ      int64
WTINT2YR    float64
SDMVPSU       int64
SDMVSTRA      int64
INDFMPIR    float64
BPXSY1      float64
BPXDI1      float64
BPXSY2      float64
BPXDI2      float64
BMXWT       float64
BMXHT       float64
BMXBMI      float64
BMXLEG      float64
BMXARML     float64
BMXARMC     float64
BMXWAIST    float64
HIQ210      float64
dtype: object

As we see here, most of the variables have floating point or integer data type.  Unlike many data sets, NHANES does not use any text values in its data.  For example, while many datasets would use text labels like "F" or "M" to denote a subject's gender, this information is represented in NHANES with integer codes.  The actual meanings of these codes can be determined from the codebooks.  For example, the variable `RIAGENDR` contains each subject's gender, with male gender coded as `1` and female gender coded as `2`.  The `RIAGENDR` variable is part of the demographics component of NHANES, so this coding can be found in the demographics codebook.

Variables like `BMXWT` which represent a quantitative measurement will typically be stored as floating point data values.

### Slicing a data set

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent cases and the columns represent variables.  One common manipulation of a data frame is to extract the data for one case or for one variable.  There are several ways to do this, as shown below.

To extract all the values for one variable, the following three approaches are equivalent ("DMDEDUC2" here is an NHANES variable containing a person's educational attainment).  In these four lines of code, we are assigning the data from one column of the data frame `da` into new variables `w`, `x`, `y`, and `z`.  The first three approaches access the variable by name.  The fourth approach accesses the variable by position (note that `DMDEDUC2` is in position 9 of the `da.columns` array shown above -- remember that Python counts starting at position zero).

In [12]:
w = da["DMDEDUC2"]
x = da.loc[:, "DMDEDUC2"]
y = da.DMDEDUC2
z = da.iloc[:, 9]  # DMDEDUC2 is in column 9


Another reason to slice a variable out of a data frame is so that we can then pass it into a function.  For example, we can find the maximum value over all `DMDEDUC2` values using any one of the following four lines of code:

In [14]:
print(da["DMDEDUC2"].max())
print(da.loc[:, "DMDEDUC2"].max())
print(da.DMDEDUC2.max())
print(da.iloc[:, 9].max())

9.0
9.0
9.0
9.0


Every value in a Python program has a type, and the type information can be obtained using Python's '`type`' function.  This can be useful, for example, if you are looking for the documentation associated with some value, but you do not know what the value's type is.  

Here we see that the variable `da` has type 'DataFrame', while one column of `da` has type 'Series'.  As noted above, a Series is a Pandas data structure for holding a single column (or row) of data.

In [15]:
print(type(da)) # The type of the variable
print(type(da.DMDEDUC2)) # The type of one column of the data frame
print(type(da.iloc[2,:])) # The type of one row of the data frame

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>


It may also be useful to slice a row (case) out of a data frame.  Just as a data frame's columns have names, the rows also have names, which are called the "index".  However many data sets do not have meaningful row names, so it is more common to extract a row of a data frame using its position.  The `iloc` method slices rows or columns from a data frame by position (counting from 0). The following line of code extracts row 3 from the data set (which is the fourth row, counting from zero).

In [16]:
x = da.iloc[3, :]
x

SEQN         83735.0
ALQ101           2.0
ALQ110           1.0
ALQ130           1.0
SMQ020           2.0
RIAGENDR         2.0
RIDAGEYR        56.0
RIDRETH1         3.0
DMDCITZN         1.0
DMDEDUC2         5.0
DMDMARTL         6.0
DMDHHSIZ         1.0
WTINT2YR    102718.0
SDMVPSU          1.0
SDMVSTRA       131.0
INDFMPIR         5.0
BPXSY1         132.0
BPXDI1          72.0
BPXSY2         134.0
BPXDI2          68.0
BMXWT          109.8
BMXHT          160.9
BMXBMI          42.4
BMXLEG          38.5
BMXARML         37.7
BMXARMC         38.3
BMXWAIST       110.1
HIQ210           2.0
Name: 3, dtype: float64

Another important data frame manipulation is to extract a contiguous block of rows or columns from the data set.  Below we slice by position, in the first case taking row positions 3 and 4 (counting from 0, which are rows 4 and 5 counting from 1), and in the second case taking columns 2, 3, and 4 (columns 3, 4, 5 if counting from 1).

### Missing values
When reading a dataset using Pandas, there is a set of values including 'NA', 'NULL', and 'NaN' that are taken by default to represent a missing value.  The full list of default missing value codes is in the '`read_csv`' documentation [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html).  This document also explains how to change the way that '`read_csv`' decides whether a variable's value is missing.

Pandas has functions called `isnull` and `notnull` that can be used to identify where the missing and non-missing values are located in a data frame.  Below we use these functions to count the number of missing and non-missing `DMDEDUC2` values.

In [17]:
print(pd.isnull(da.DMDEDUC2).sum())
print(pd.notnull(da.DMDEDUC2).sum())

261
5474


As an aside, note that there may be a variety of distinct forms of missingness in a variable, and in some cases it is important to keep these values distinct.  For example, in case of the DMDEDUC2 variable, in addition to the blank or NA values that Pandas considers to be missing, three people responded "don't know" (code value 9).  In many analyses, the "don't know" values will also be treated as missing, but at this point we are considering "don't know" to be a distinct category of observed response.